power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/handle.hpp

```diff
@@ -0,0 +1,100 @@
+    const char* what() const noexcept override { return message_.c_str(); }
+
+  private:
+    std::string message_;
+};
+
+class PowerGridRegularError : public PowerGridError {
+  public:
+    using PowerGridError::PowerGridError;
+    static constexpr Idx error_code() { return PGM_regular_error; }
+};
+
+class PowerGridSerializationError : public PowerGridError {
+  public:
+    using PowerGridError::PowerGridError;
+    static constexpr Idx error_code() { return PGM_serialization_error; }
+};
+
+class PowerGridBatchError : public PowerGridError {
+  public:
+    struct FailedScenario {
+        Idx scenario;
+        std::string error_message;
+    };
+
+    PowerGridBatchError(std::string const& message, std::vector<FailedScenario> failed_scenarios_c)
+        : PowerGridError{message}, failed_scenarios_{std::move(failed_scenarios_c)} {}
+    static constexpr Idx error_code() { return PGM_batch_error; }
+    std::vector<FailedScenario> const& failed_scenarios() const { return failed_scenarios_; }
+
+  private:
+    std::vector<FailedScenario> failed_scenarios_;
+};
+
+class Handle {
+  public:
+    Handle() : handle_{PGM_create_handle()} {}
+
+    RawHandle* get() const { return handle_.get(); }
+
+    static void check_error(Handle const& handle) {
+        RawHandle* handle_ptr = handle.get();
+        Idx error_code = PGM_error_code(handle_ptr);
+        std::string error_message = error_code == PGM_no_error ? "" : PGM_error_message(handle_ptr);
+        switch (error_code) {
+        case PGM_no_error:
+            return;
+        case PGM_regular_error:
+            throw PowerGridRegularError{error_message};
+        case PGM_batch_error: {
+            Idx const n_failed_scenarios = PGM_n_failed_scenarios(handle_ptr);
+            std::vector<PowerGridBatchError::FailedScenario> failed_scenarios(n_failed_scenarios);
+            auto const failed_scenario_seqs = PGM_failed_scenarios(handle_ptr);
+            auto const failed_scenario_messages = PGM_batch_errors(handle_ptr);
+            for (Idx i = 0; i < n_failed_scenarios; ++i) {
+                failed_scenarios[i] =
+                    PowerGridBatchError::FailedScenario{failed_scenario_seqs[i], failed_scenario_messages[i]};
+            }
+            throw PowerGridBatchError{error_message, std::move(failed_scenarios)};
+        }
+        case PGM_serialization_error:
+            throw PowerGridSerializationError{error_message};
+        default:
+            throw PowerGridError{error_message};
+        }
+    }
+    void check_error() const { check_error(*this); }
+
+    static void clear_error(Handle& handle) { PGM_clear_error(handle.get()); }
+    void clear_error() { clear_error(*this); }
+
+    template <typename Func, typename... Args> auto call_with(Func&& func, Args&&... args) const {
+        auto result = std::forward<Func>(func)(get(), std::forward<Args>(args)...);
+        check_error();
+        clear_error();
```

Comment:
Hmmm, this will not work as intended. If `check_error` throws, the `clear_error` will never be called. So the error still stays.

We need a RAII principle to make sure the `check_error` is called. Define a `unique_ptr` to wrap the `PGM_clear_error`. When it goes out of scope, the `PGM_clear_error` will be called.

```suggestion
        detail::UniquePtr<RawHandle, PGM_clear_error> clear_handle_error{handle.get()};
        check_error();
```

Reply:
Or alternatively just get the message, store temporarily in local variable, call clear error, and then throw exception with moved message.

That is ok because it's fallback code and performance is no issue. Also, cross-dll is more expensive than moving a simple string

Reply:
Or we can just `handle.clear_error()` before every `throw` statement within each `switch` `case`, because at that point the `handle` does not really matter anymore.